### PR TITLE
Implement ensemble classifiers

### LINKED ISSUE #305: Super-confident predictions
Different classifiers may capture different relations in the data. We can join what each classifier learns by creating ensembles of them. 

- [x] Use sklearn's VotingClassifier instead of our implementation - #366 
- [x] Implement a Gating ensemble - #356 
- [x] Implement a Stacking ensemble - #367 

Results will then be presented in a separate task

- [x]  Analyze performance of ensemble methods - #359

### LINKED ISSUE #356: Implement a Gating ensemble
see comments in #348

### LINKED ISSUE #367: Implement a Stacking ensemble
see comments in #348

### PR DESCRIPTION
**Pull request by tupini07**
This PR implements the remaining ensemble classifiers (closes #305):

- clean up VotingClassifier (#366)
- Gated ensemble (closes #356)
- Stack ensemble (closes #367)

For the record, the results (evaluation and linking) of these classifiers have been posted in issue #359

It also introduces some refactoring of the classifier module (reordering of the classifier definitions)

Finally, the scripts used to extract the _reports_ posted in #359 have been included in the `scripts/linker` folder.

### CODE REVIEW

File: soweego/linker/classifiers.py [hunk 1]
```diff
@@ -441,3 +396,184 @@ def _create_model(
...
+
+    def __repr__(self):
+        return (
+            f'{self.__class__.__name__}('
+            f'num_folds={self.num_folds}, '
+            f'meta_layer={self.meta_layer}) '
+        )
+
+
+class StackedEnsambleClassifier(SKLearnAdapter, BaseClassifier):
```
Comment (marfox) [inline soweego/linker/classifiers.py]: _Stacked_, _Stacking_, or _Stack_? _Ensemble_, not _Ensamble_

File: soweego/linker/classifiers.py [hunk 2]
```diff
@@ -441,3 +396,184 @@ def _create_model(
...
+    """Ensemble of stacked classifiers, meaning that classifiers are arranged in layers
```
Comment (marfox) [inline soweego/linker/classifiers.py]: Please make the docstring standard.

File: soweego/linker/classifiers.py [hunk 3]
```diff
@@ -441,3 +396,184 @@ def _create_model(
...
+            estimators = []
+            for clf in constants.CLASSIFIERS_FOR_ENSEMBLE:
+                model = utils.init_model(
+                    clf, num_features=num_features, **kwargs
+                )
+
+                estimators.append((clf, model.kernel))
+            return estimators
+
+        self.kernel = SuperLearner(verbose=2, n_jobs=1, folds=self.num_folds)
```
Comment (marfox) [inline soweego/linker/classifiers.py]: Same comment as per the gate classifier

File: soweego/commons/constants.py [hunk 4]
```diff
@@ -280,6 +280,8 @@
     'single_layer_perceptron': keys.SINGLE_LAYER_PERCEPTRON,
     'multi_layer_perceptron': keys.MULTI_LAYER_PERCEPTRON,
     'voting_classifier': keys.VOTING_CLASSIFIER,
+    'gate_classifier': keys.GATE_CLASSIFIER,
```
[the remark this answers is not in the document]
Comment (tupini07) [reply soweego/commons/constants.py]: Stacked. I've changed it in a87d0a3a4a9da9c7d9a5c94da3c071b73bde382a

File: soweego/commons/constants.py [hunk 5]
```diff
@@ -402,8 +406,16 @@
     ),
 }
 
+# Parameters for ensemble
 VOTE_CLASSIFIER_PARAMS = {"voting": "hard"}
```
[the remark this answers is not in the document]
Comment (tupini07) [reply soweego/commons/constants.py]: good point! Done in 1996634a1b58954db3186c82fc3402ab8c32f37e

File: soweego/linker/classifiers.py [hunk 6]
```diff
@@ -441,3 +396,184 @@ def _create_model(
         model.compile(optimizer=optimizer, loss=self.loss, metrics=self.metrics)
 
         return model
+
+
+class VoteClassifier(SKLearnAdapter, BaseClassifier):
```
[the remark this answers is not in the document]
Comment (tupini07) [reply soweego/linker/classifiers.py]: Done in 535f343dddd9876cc0226b8a8c3d0f55fca24dc9 !

File: soweego/linker/classifiers.py [hunk 7]
```diff
@@ -441,3 +396,184 @@ def _create_model(
...
+    """Basic ensemble classifier which chooses the correct prediction by
+    using majority voting (aka 'hard' voting) or chooses the label which has the
```
[the remark this answers is not in the document]
Comment (tupini07) [reply soweego/linker/classifiers.py]: done in a42c5eb4380b1cebea24383fb8acf537625fd620

File: soweego/linker/classifiers.py [hunk 8]
```diff
@@ -441,3 +396,184 @@ def _create_model(
...
+            # get only the probability that pairs are a match
+            classifications = self.kernel.predict_proba(feature_vectors)[:, 1]
+
+        return pd.Series(classifications, index=feature_vectors.index)
+
+    def __repr__(self):
+        return f'{self.kernel}'
+
+
+class GateEnsambleClassifier(SKLearnAdapter, BaseClassifier):
```
[the remark this answers is not in the document]
Comment (tupini07) [reply soweego/linker/classifiers.py]: Fixed both, a87d0a3a4a9da9c7d9a5c94da3c071b73bde382a and 5acd67ddba95437cdcc194d5efcb9e8980080dc2 respectively

File: soweego/linker/classifiers.py (hunk 1 above)
Comment (tupini07) [reply soweego/linker/classifiers.py]: Fixed both, a87d0a3 and 5acd67d respectively

**Comment by tupini07** [comment]
@marfox thanks for the comprehensive review! I still haven't finished fixing all the issues you mentioned. I just wanted to let you know that I just included a new feature in this PR: now classifiers which use `mlens` under the hood (_GatedClassifier_ and _StackedClassigier_) return the probability of a match instead of just whether it was a match or not. This has been done in 5318d5f

File: docs/conf.py [hunk 9]
```diff
@@ -92,4 +92,5 @@
     'requests': ('https://2.python-requests.org/en/stable/', None),
```
Comment (marfox) [inline docs/conf.py]:
```suggestion
     'requests': ('https://requests.readthedocs.io/en/stable/', None),
```
Just found a fresh URL for `requests`

File: soweego/linker/classifiers.py [hunk 10]
```diff
@@ -37,12 +38,152 @@
...
+    def _predict(self, features) -> np.ndarray:
+        prediction = super(_MLensAdapter, self)._predict(features)
+        self._check_correct_pred_shape(prediction)
+
+        prediction = prediction[:, 1]
+
+        # mlens `predict` method returns probabilities. Since we're
+        # dealing with a binary classification problem we just get the
+        # probabilities for the positive case and round them to [0,1].
+        prediction = np.array(list(round(x) for x in prediction))
```
Comment (marfox) [inline soweego/linker/classifiers.py]: Could you explain what's the exact goal of this decision? Does `round them to [0,1]` mean that we're dropping the exact probability scores?

Comment (tupini07) [reply soweego/linker/classifiers.py]: Thanks @marfox you raise a good point. However, as you can see I've added two methods: `prob` which returns the raw predictions, and `_predict` which returns the predicted label (that is why I'm using `round`). 

To clarify, `_predict` is used by `recordlinkage` during the evaluation, since it needs the final class labels to calculate the metrics. `prob` is used the rest of the time (in `train` and `link`)

File: soweego/linker/classifiers.py [hunk 11]
```diff
@@ -441,3 +396,184 @@ def _create_model(
...
+            classifications = self.kernel.predict_proba(feature_vectors)[:, 1]
+
+        return pd.Series(classifications, index=feature_vectors.index)
+
+    def __repr__(self):
+        return f'{self.kernel}'
+
+
+class GateEnsambleClassifier(SKLearnAdapter, BaseClassifier):
+    """Ensemble of classifiers, whose predictions are joined by using
```
[the remark this answers is not in the document]
Comment (tupini07) [reply soweego/linker/classifiers.py]: Done in 0e53b81fb342a6ac807c6769d43db5190a1c8530

File: soweego/linker/classifiers.py [hunk 12]
```diff
@@ -441,3 +396,184 @@ def _create_model(
...
+
+        estimators = []
+        for clf in constants.CLASSIFIERS_FOR_ENSEMBLE:
+            model = utils.init_model(
+                clf, num_features=self.num_features, **kwargs
+            )
+
+            estimators.append((clf, model.kernel))
+
+        self.kernel = SuperLearner(verbose=2, n_jobs=1, folds=self.num_folds)
```
[the remark this answers is not in the document]
Comment (tupini07) [reply soweego/linker/classifiers.py]: The `verbose=2` gives us log messages when training/classifying with each layer. 

And yes, this kind of ensemble has a great potential for parallelization. Specifically, each model can be trained on a separate thread. Sadly, Keras has trouble moving the ownership of the memory containing the weights of the network from thread to thread, so using more than 1 job will make Keras fail. The performance with only one thread is not so bad, so I didn't put much effort into solving this issue :)

File: soweego/linker/classifiers.py (hunk 2 above)
Comment (tupini07) [reply soweego/linker/classifiers.py]: Done in b5571120c0b584a8f3120c89ad1923f6c812532d

File: soweego/linker/classifiers.py (hunk 3 above)
Comment (tupini07) [reply soweego/linker/classifiers.py]: Same answer as given for `GatedEnsembleClassifier` :smile:

File: scripts/linker/analyze_classification_links.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
+"""
+This script creates a summary and graphs derived from the classification results.
+
+It needs to be executed inside the directory which contains the classification links files
+"""
+
+from glob import glob
```
[the remark this answers is not in the document]
Comment (tupini07) [reply scripts/linker/analyze_classification_links.py]: good idea! Added in c8f36eea1bd9572885cac5d9c970ba3f889d3151

File: scripts/linker/analyze_classification_links.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+all_links = pd.DataFrame(columns=[
+    "Catalog",
+    "Entity",
+    "Model",
+    "QID",
+    "TID",
+    "Prediction"
+])
+
+non_binary_classifiers = [
```
[the remark this answers is not in the document]
Comment (tupini07) [reply scripts/linker/analyze_classification_links.py]: yes it might be a bit confusing. I've renamed this list to _continuous_classifiers_ in b7922195e04a1390c74413c9de3b5cc2a53e3543 . The distinction is that these classifiers give as a result a prediction which is a continuous value (instead of just the label `[0,1]`)

File: scripts/linker/analyze_classification_links.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+print(df.to_csv(index=False))
+
+###############
+# Get summaries of the statistics and print them
+summaries = []
+for _, gg in df.groupby('Model'):
+    # F1.Mean    F1.STD Prec.Mean  Prec.STD Recall.Mean Recall.STD
+    summaries.append({
+        "Model": gg["Model"].values[0],
+        "Average Mean": "%.6f" % gg['Mean'].apply(lambda x: float(x)).mean(),
```
[the remark this answers is not in the document]
Comment (tupini07) [reply scripts/linker/analyze_classification_links.py]: 1. Very good point! Fixed in ed3a7bb7cd26953c535f70346a16a518f08bdf26
2. I searched a bit online and couldn't find any difference :open_mouth: maybe I didn't search well enough :P . The idea is for this value to be the `mean` value of the `mean` results

File: scripts/linker/analyze_classification_links.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+    }, ignore_index=True)
+
+# print all files' statistics
+df = df.sort_values(by=['Model', 'Catalog'])
+print(df.to_csv(index=False))
+
+###############
+# Get summaries of the statistics and print them
+summaries = []
+for _, gg in df.groupby('Model'):
```
[the remark this answers is not in the document]
Comment (tupini07) [reply scripts/linker/analyze_classification_links.py]: Yes you're right. But we would still need to process the result of `describe` to get the information we're interested in presenting (since `describe` yields much more information than what we're interested in). 

If you think it would be cleaner to use `df.describe()` anyway then let me know and I can change it.

[the remark this answers is not in the document]
Comment (marfox) [reply scripts/linker/analyze_classification_links.py]: fair enough

File: soweego/linker/classifiers.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (marfox) [reply soweego/linker/classifiers.py]: That's too bad, thanks for the explanation.

File: scripts/linker/analyze_classification_links.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (marfox) [reply scripts/linker/analyze_classification_links.py]: fair enough
